src/Grpc.AspNetCore.Server/GrpcEndpointRouteBuilderExtensions.cs

```diff
@@ -20,60 +20,94 @@
+        {
+            if (builder == null)
+            {
+                throw new ArgumentNullException(nameof(builder));
+            }
+
             ValidateServicesRegistered(builder.ServiceProvider);
 
-            var service = typeof(TService);
+            var options = new GrpcBindingOptions<TService>();
+            options.BindAction = ReflectionBind<TService>;
+            options.InvokerFactory = new ReflectionMethodInvokerFactory<TService>();
+
+            configureOptions?.Invoke(options);
+
+            var callHandlerFactory = builder.ServiceProvider.GetRequiredService<ServerCallHandlerFactory<TService>>();
+            var serviceBinder = new GrpcServiceBinder<TService>(builder, callHandlerFactory, options.InvokerFactory);
 
             try
             {
-                // TService is an implementation of the gRPC service. It ultimately derives from Foo.TServiceBase base class.
-                // We need to access the static BindService method on Foo which implicitly derives from Object.
-                var baseType = service.BaseType;
-                while (baseType?.BaseType?.BaseType != null)
-                {
-                    baseType = baseType.BaseType;
-                }
-
-                // We need to call Foo.BindService from the declaring type.
-                var declaringType = baseType?.DeclaringType;
+                options.BindAction(serviceBinder, null);
+            }
+            catch (Exception ex)
+            {
+                throw new InvalidOperationException($"Error binding gRPC service '{typeof(TService).Name}'.", ex);
+            }
 
-                // The method we want to call is public static void BindService(ServiceBinderBase, BaseType)
-                var bindService = declaringType?.GetMethod("BindService", new[] { typeof(ServiceBinderBase), baseType });
+            return new CompositeEndpointConventionBuilder(serviceBinder.EndpointConventionBuilders);
+        }
 
-                if (bindService == null)
-                {
-                    throw new InvalidOperationException($"Cannot locate BindService(ServiceBinderBase, ServiceBase) method for the current service type: {service.FullName}. The type must be an implementation of a gRPC service.");
-                }
+        private static void ReflectionBind<TService>(ServiceBinderBase binder, TService _)
+        {
+            var service = typeof(TService);
 
-                var callHandlerFactory = builder.ServiceProvider.GetRequiredService<ServerCallHandlerFactory<TService>>();
+            // TService is an implementation of the gRPC service. It ultimately derives from Foo.TServiceBase base class.
+            // We need to access the static BindService method on Foo which implicitly derives from Object.
+            var baseType = service.BaseType;
+            while (baseType?.BaseType?.BaseType != null)
```

Comment:
I know this logic isn't new in this PR, but it cannot be correct to jump over two base types at a time. It looks like this code is trying to get the type that derives from System.Object (e.g. Greet.Greeter+GreeterBase) instead of the object type itself. The problem is that this only works if there is an even number of base types. So it works for:

```
GreeterService -> Greet.Greeter+GreeterBase -> System.Object -> null
```

But it breaks for:

```
GreeterService -> IntermediateDerivedType -> Greet.Greeter+GreeterBase -> System.Object -> null
```

In the second case, baseType would be System.Object instead of Greet.Greeter+GreeterBase as expected when the loop ends.

Reply:
🤷‍♂️ 

I think it has been copy and pasted through the ages. I'll check it out

Reply:
As far as I can tell it never did anything.